API Update

marc-tonsen opened the pull request:
Changes based on naming convention formulated [here](https://www.notion.so/pupillabs/PL-Libraries-API-Naming-Guide-Proposal-1fc9d188973c802db430f85c16f4cc17).

dourvaris commented on src/pupil_labs/neon_recording/timeseries/saccades.py:
```diff
@@ -0,0 +1,104 @@
+        )
+        data = data[data["event_type"] == 0]
+        data = data[["start_timestamp_ns", "end_timestamp_ns", "start_gaze_x", "start_gaze_y", "end_gaze_x", "end_gaze_y", "amplitude_angle_deg", "mean_velocity", "max_velocity"]]
+        data.dtype.names = (
```
This will break when data contains more (or less) fields

dourvaris reviewed (changes requested):
Before we can think of merging this we will need a full test suite on the new api.

Currently the following below fail on latest app version recording:

```python
rec = nr.open('/rec').eyeball.center_left # KeyError: 'eyeball_center_left_x'
rec = nr.open('/rec').eyelid # KeyError: 'eyelid_angle_top_left'
```

marc-tonsen commented:
@dourvaris I updated things to be compatible with current production recordings and also fixed the tests. The current set of test recordings does not contain eyelid data though.

Beyond the existing tests and adding recordings with eyelid data, what else do you think is needed?

domstoppable commented on src/pupil_labs/neon_recording/timeseries/fixations.py:
```diff
@@ -0,0 +1,107 @@
+    stop_time = fields[np.int64]("stop_time")  # type:ignore
+    "Stop timestamp of fixation."
+
+    start_gaze = fields[np.float32]([
```
Should `start/stop/mean_gaze*` names be `start_gaze_point` instead?

Or rather than `start_time` and `start_gaze`, we could have `start_gaze.time` and `start_gaze.point`?

I'm ok with the API as it is, but I can see arguments for the options above

marc-tonsen replied:
I agree with `start_gaze_point`. That is more consistent!

We had been considering additional layers of hierarchy kinda like you suggest with `start_gaze.time`. Similarly one could do something like `rec.pupil.left.diameter`. The conclusion was that a deeper hierarchy like this quickly becomes difficult to navigate though and the flat one that is currently used is thus preferred, even though some attribute lists become a bit long and naming slightly redundant. Unless we find good new arguments, I'd vote for sticking with what we currently have.

marc-tonsen commented on src/pupil_labs/neon_recording/timeseries/fixations.py:
```diff
@@ -0,0 +1,107 @@
+                ("max_velocity", "float32"),
+            ]),
+        )
+        data = data[data["event_type"] == 1]
```
The arguments for the present approach are:
- saccade data is not of interest to the majority of our users, while fixation data is very widely used. Thus, grouping the two together would introduce unnecessary complexity for most users and the ned to filter down the data.
- we are providing different metrics for the two, i.e. amplitude for saccades, which does not make sense for fixations. This makes the dtype more complicated. When e.g. exporting to a DataFrame one would have lots of nan values in it. When filtering the data down to just fixations, one would also have to drop the useless saccade-related columns.

Given these arguments, would you still prefer to group these together?

domstoppable replied:
Not sure how I missed that 😅

Are we missing `amplitude_pixels`?

marc-tonsen replied:
Yes, but it was intentionally omitted. Representing spatial values relating to the scene camera in pixels is not ideal, since the pixel space is distorted by the camera projection. Instead, any such values should be given in angles in theory.

The idea was to make this switch now and either change or drop values given in pixels. The amplitude was given in both pixels and degrees, so one is dropped the other one stays. The max and mean velocities should also change to deg / s although I see the docstring was not yet changed accordingly.

Thinking about this, we previously also said that all angles should be given in radians rather than degrees, so maybe we should use radians for these here too.

domstoppable replied:
angles > pixels I generally agree with, but I don't see the problem here given that
1. The provided scene camera images are distorted, so I'd expect everything in pixel-space to be distorted too. These data align
2. It seems a bit odd to hide a value that's already computed. If a user wants this value, they'd have to re-compute it themselves

Having said that, is a saccade amplitude measured in pixels ever a useful metric? I doubt it, but don't really know. If it's not useful, then including it would just be noise

marc-tonsen replied:
Correct, the saccades would technically be correct and incorporate the distortion. But this representation is not useful, especially if the angular version is available.

The camera distortion is an issue because a saccade in the center of the camera's FOV may have the same amplitude measured in pixels as another saccade in the corner of the camera's FOV, but the actual movement of the eyeballs associated with each saccade may differ a lot. The angular representation makes this visible.
